### PR TITLE
Improve ec encode performance

### PR DESCRIPTION
**Pull request by sek2022**
# What problem are we solving?
1.Master Ui improve
2.Ec encode is too slow！

# How are we solving the problem?
1. Processing multiple volumes concurrently.
2. Encode is written directly to the target server, No need to write to the encoding server.
3. Fix some bug.
4. Add Rack Statistics for master UI.

截屏2024-10-10 16 49 40

# How is the PR tested?

Add parameter: concurrentNumber, The number of volumes to be processed concurrently.

`
weed shell

ec.encode -fullPercent=95 -quietFor=5s -concurrentNumber=1

ec.encode -fullPercent=95 -quietFor=5s -concurrentNumber=3
...
ec.encode -fullPercent=95 -quietFor=5s -concurrentNumber=12

ec.rebuild -force
ec.balance -force
...

`

# Checks
- [x] I have added unit tests if possible.
- [x] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/server/volume_grpc_upload_test.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"fmt"
+	"os"
+	"testing"
+)
+
+func TestVolumeServer_UploadFile(t *testing.T) {
+	flags := os.O_WRONLY | os.O_CREATE | os.O_TRUNC
+	baseFileName := "/Users/ym/seaweedfs/test/t1.vif"
```
Comment (chrislusf) [inline weed/server/volume_grpc_upload_test.go]: need to fix the path here.

**Comment by sek2022** [comment]
ok

「Chris Lu ***@***.***>」在 2024年10月10日 週四，下午4:58 寫道：

**Comment by chrislusf** [comment]
> # How is the PR tested?
> Already tested

Please add more details on this.

**Comment by sek2022** [comment]
> > # How is the PR tested?
> > Already tested
> 
> Please add more details on this.

Thank you！I have modified！

File: weed/server/volume_grpc_upload_test.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 
 func TestVolumeServer_UploadFile(t *testing.T) {
 	flags := os.O_WRONLY | os.O_CREATE | os.O_TRUNC
-	baseFileName := "~/t1.vif"
+	baseFileName := "/tmp/test.vif"
```
Comment (chrislusf) [inline weed/server/volume_grpc_upload_test.go]: use `os.TempDir()`

File: weed/pb/volume_server.proto [hunk 3]
```diff
@@ -350,9 +364,14 @@ message VolumeTailReceiverRequest {
 message VolumeTailReceiverResponse {
 }
 
+message EcIds {
```
Comment (chrislusf) [inline weed/pb/volume_server.proto]: please change `EcIds` to `EcShardIds`

File: weed/shell/command_ec_common_test.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"fmt"
+	"testing"
+)
+
+func Test_ceilDivide(t *testing.T) {
+	//3
+	result1 := ceilDivide(14, 5)
+	fmt.Println(result1)
```
Comment (chrislusf) [inline weed/shell/command_ec_common_test.go]: assert

File: weed/shell/command_ec_encode_test.go [hunk 5]
```diff
@@ -29,3 +35,127 @@ func TestEcDistribution(t *testing.T) {
...
+	fmt.Printf("volumeChooseLocationMap: %v \n", volumeChooseLocationMap)
+
+	var wg sync.WaitGroup
+	wg.Add(len(volumeIds))
+	for _, vid := range volumeIds {
+		locations := volumeLocationsMap[vid]
+		chooseLoc := volumeChooseLocationMap[vid]
+		go func() {
+			defer wg.Done()
+			time.Sleep(time.Second * 1)
```
Comment (chrislusf) [inline weed/shell/command_ec_encode_test.go]: use proper locking instead.

File: weed/storage/volume_write.go [hunk 6]
```diff
@@ -80,16 +80,18 @@ func (v *Volume) Destroy(onlyEmpty bool) (err error) {
...
 	v.doClose()
-	removeVolumeFiles(v.DataFileName())
-	removeVolumeFiles(v.IndexFileName())
+	removeVolumeFiles(v.DataFileName(), afterEc)
+	removeVolumeFiles(v.IndexFileName(), afterEc)
 	return
 }
 
-func removeVolumeFiles(filename string) {
+func removeVolumeFiles(filename string, afterEc bool) {
```
Comment (chrislusf) [inline weed/storage/volume_write.go]: `afterEc` only shows when. Please change to a more meaningful name for what it does.

File: weed/storage/volume_info/volume_info.go [hunk 7]
```diff
@@ -82,3 +82,48 @@ func SaveVolumeInfo(fileName string, volumeInfo *volume_server_pb.VolumeInfo) er
...
+			Collection:  collection,
+			VolumeId:    volumeId,
+			Ext:         ext,
+			FileContent: text,
+		})
+		if clientErr != nil {
+			return fmt.Errorf("client send error ext: %s: %v", ext, clientErr)
+		}
+	}
+	//if err := util.WriteFile(fileName, text, 0644); err != nil {
```
Comment (chrislusf) [inline weed/storage/volume_info/volume_info.go]: clean up unused code

File: weed/shell/command_ec_encode.go [hunk 8]
```diff
@@ -98,76 +101,205 @@ func (c *commandEcEncode) Do(args []string, commandEnv *CommandEnv, writer io.Wr
...
+		return errors[0]
 	}
 
 	return nil
 }
 
-func doEcEncode(commandEnv *CommandEnv, collection string, vid needle.VolumeId, parallelCopy bool) (err error) {
-	if !commandEnv.isLocked() {
-		return fmt.Errorf("lock is lost")
+// 递归函数用于获取数组的排列组合
```
Comment (chrislusf) [inline weed/shell/command_ec_encode.go]: use english comments

File: weed/shell/command_ec_encode.go [hunk 9]
```diff
@@ -98,76 +101,205 @@ func (c *commandEcEncode) Do(args []string, commandEnv *CommandEnv, writer io.Wr
...
+func permute(arr []int, n int, result *[][]int) {
```
Comment (chrislusf) [inline weed/shell/command_ec_encode.go:163]: result does not need to be a pointer to an array. Use a return value.

File: weed/shell/command_ec_encode.go [hunk 10]
```diff
@@ -69,6 +71,7 @@ func (c *commandEcEncode) Do(args []string, commandEnv *CommandEnv, writer io.Wr
 	quietPeriod := encodeCommand.Duration("quietFor", time.Hour, "select volumes without no writes for this period")
 	parallelCopy := encodeCommand.Bool("parallelCopy", true, "copy shards in parallel")
 	forceChanges := encodeCommand.Bool("force", false, "force the encoding even if the cluster has less than recommended 4 nodes")
+	concurrentNumber := encodeCommand.Int("concurrentNumber", 3, "limit total concurrent ec.encode volume number (default 3)")
```
Comment (chrislusf) [inline weed/shell/command_ec_encode.go]: change `concurrentNumber` to `concurrency`

File: weed/shell/command_ec_encode.go [hunk 11]
```diff
@@ -98,76 +101,205 @@ func (c *commandEcEncode) Do(args []string, commandEnv *CommandEnv, writer io.Wr
...
+		return ""
+	}
+	return parts[0]
+}
 
+func doEcEncode(commandEnv *CommandEnv, collection string, vid needle.VolumeId, locations []wdclient.Location, chooseLoc wdclient.Location, parallelCopy bool) (err error) {
+	if !commandEnv.isLocked() {
+		return fmt.Errorf("lock is lost")
+	}
+	if len(locations) == 0 {
```
Comment (chrislusf) [inline weed/shell/command_ec_encode.go:279]: why this extra lookup is needed?

File: weed/shell/command_ec_encode.go [hunk 12]
```diff
@@ -98,76 +101,205 @@ func (c *commandEcEncode) Do(args []string, commandEnv *CommandEnv, writer io.Wr
...
+	volumeLocationsMap, volumeChooseLocationMap := chooseLocationForVolumes(commandEnv, processVolumeIds)
+
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var errors = make([]error, 0)
+	wg.Add(len(processVolumeIds))
+
+	for _, vid := range processVolumeIds {
+		locations := volumeLocationsMap[vid]
+		chooseLoc := volumeChooseLocationMap[vid]
```
Comment (chrislusf) [inline weed/shell/command_ec_encode.go]: how to prevent multiple ec processes running on the same volume server?

**Comment by chrislusf** [comment]
The PR is too big to merge. Please split into multiple smaller PRs.

**Pull request closed without merging**